src/main/cpp/subsystems/SwerveDrive.cpp

```diff
@@ -0,0 +1,559 @@
+    //TODO convert `turn` to rad/sec
+    //TODO convert drive and strafe to m/s I think
+    // because tracklength/width are in meters
+
+    // copied from ChiefDelphi thread
+    //TODO post link here
+    double A = strafe - turn * tracklength/2;
+    double B = strafe + turn * tracklength/2;
+    double C = drive - turn * trackwidth/2;
+    double D = drive + turn * trackwidth/2;
+
+    // speed
+    double left_front_speed = sqrt(pow(B, 2) + pow(D, 2));
+    double left_back_speed = sqrt(pow(A, 2) + pow(D, 2));
+    double right_front_speed = sqrt(pow(B, 2) + pow(C, 2));
+    double right_back_speed = sqrt(pow(A, 2) + pow(C, 2));
+
+
+    // turn
+    double left_front_turn = atan2(B, D)  *  180.0/pi;
+    double left_back_turn = atan2(A, D)  *  180.0/pi;
+    double right_front_turn = atan2(B, C)  *  180.0/pi;
+    double right_back_turn = atan2(A, C)  *  180.0/pi;
+
+    // std::cout << left_front_turn << std::endl;
+
+    if (left_front_turn < 0) {
+        left_front_turn += 360;
+    }
+
+    if (left_back_turn < 0) {
+        left_back_turn += 360;
+    }
+
+    if (right_front_turn < 0) {
+        right_front_turn += 360;
+    }
+    
+    if (right_back_turn < 0) {
+        right_back_turn += 360;
+    }
+    
+    OKC_CALL(this->left_front_module->SetAngle(left_front_turn));
+    OKC_CALL(this->left_back_module->SetAngle(left_back_turn));
+    OKC_CALL(this->right_front_module->SetAngle(right_front_turn));
+    OKC_CALL(this->right_back_module->SetAngle(right_back_turn));
+
+    OKC_CALL(this->left_front_module->GetSteerOutput(&this->interface_->left_front_steer_motor_output));
+    OKC_CALL(this->left_back_module->GetSteerOutput(&this->interface_->left_back_steer_motor_output));
+    OKC_CALL(this->right_front_module->GetSteerOutput(&this->interface_->right_front_steer_motor_output));
+    OKC_CALL(this->right_back_module->GetSteerOutput(&this->interface_->right_back_steer_motor_output));
+
+    this->interface_->left_front_drive_motor_output = left_front_speed;
+    this->interface_->left_back_drive_motor_output = left_back_speed;
+    this->interface_->right_front_drive_motor_output = right_front_speed;
+    this->interface_->right_back_drive_motor_output = right_back_speed;
+
+
+    return true;
+}
+
+
+
+bool SwerveDrive::InitAuto(frc::Pose2d pos, bool keep_heading) {
+    this->auto_lock_heading = keep_heading;
+
+    //  1. figure out angle between here and there
+    this->heading_to_goal = atan((position->X().value() - pos.X().value()) / (position->Y().value() - pos.Y().value()));
+
+    //  2. figure out distance
+    this->distance_to_goal = sqrt(pow(position->X().value() + pos.X().value(), 2) + pow(position->Y().value() + pos.Y().value(), 2));
+
+    // set heading PID setpoint
+    this->heading_pid->SetSetpoint(this->heading_to_goal);
+
+    // initialization has passed, go ahead and start the autonomous
+    auto_state = INIT;
+
+    return true;
+}
+
+bool SwerveDrive::RunAuto() {
```

Comment:
This could totally be run in Periodic and then the command would just set something like `mode = PoseFollowAuto` and then this just runs